geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java

```diff
@@ -1282,14 +1282,18 @@ public boolean invalidate(EntryEventImpl event, boolean invokeCallbacks, boolean
     DiskRegion dr = owner.getDiskRegion();
     boolean ownerIsInitialized = owner.isInitialized();
 
+    // lock before waitForIndexInit so that we should wait
+    // till a concurrent clear to finish
+    lockForCacheModification(owner, event);
```

Comment:
I am not sure I understand why the code change here can fix the hang, and not causing a different type of hang. I do see the two-line comment above. But it does not answer my question. It would be helpful for the reviewers to understand it, if there is some explanation of root cause analysis and the fix in the pull request description or in the JIRA. It will also help for future code maintenance.

Reply:
It would bee nice to run a full regression with this change to see if it will cause another hang. Right now I don't think I am fully qualified to explain to you on that.

Reply:
@jchen21 
There is locking mechanism to co-ordinate between cache modification (index maintenance) and index initialization. During index initialization all the cache operations are blocked.
As part of index management; clear invokes index re-initialization. 
Also, during clear no cache operations are allowed; this is co-ordinated through lockForCacheModification (version lock).
The hang is observed due to:
Thread1 calling update, which blocks index initialization.
Thread2 calling clear, takes the cache modification lock, this blocks all cache operation; and clear triggers index re-initialization; this is now blocked for thread1 to finish update. But thread1 is waiting for cache modification lock which is held by thread2; resulting in dead-lock.

Reply:
Thanks @agingade ! This is exactly what I was looking for. Your comment is automatically logged in the JIRA, so we have the context in JIRA now. Looking at the code change of #44431 and #45699, where `waitForIndexInit()` and `lockForCacheModification()` are introduced respectively, I don't see any side effect of swapping the order of them. So I expect it should be good to fix this hang, as long as Darrel's concern is addressed.